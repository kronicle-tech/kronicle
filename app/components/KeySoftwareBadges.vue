<template>
  <div v-if="items && items.length > 0"
       class="h5"
  >
    <b-badge v-for="(item, itemIndex) in items"
             :key="itemIndex"
             variant="info"
             class="mr-2 mb-2"
    >
      {{ item.name }} {{ item.versions }}
    </b-badge>
  </div>
</template>

<script lang="ts">
import Vue, { PropType } from 'vue'
import { KeySoftware } from "~/types/kronicle-service";

interface Item {
  name: String
  versions: String
}

export default Vue.extend({
  props: {
    keySoftware: {
      type: Array as PropType<KeySoftware[]>,
      default: undefined,
    },
  },
  computed: {
    items(): Item[] {
      return (this.keySoftware || [] as KeySoftware[])
        .map(entry => ({
          name: entry.name,
          versions: entry.versions.sort().join(', '),
        }) as Item)
    }
  }
})
</script>
