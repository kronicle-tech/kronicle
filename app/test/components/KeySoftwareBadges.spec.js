import { mount } from '@vue/test-utils'
import KeySoftwareBadges from '@/components/KeySoftwareBadges.vue'

describe('KeySoftwareBadges', () => {
  let propsData
  let wrapper
  const createWrapper = () => {
    wrapper = mount(KeySoftwareBadges, { propsData })
  }

  beforeEach(() => {
    propsData = {}
  })

  afterEach(() => {
    wrapper.destroy()
    wrapper = null
  })

  describe('when keySoftware prop is not set', () => {
    test('renders nothing', () => {
      createWrapper()
      expect(wrapper.html()).toEqual(``)
    })
  })

  describe('when keySoftware prop is set to an empty array', () => {
    beforeEach(() => {
      propsData.keySoftware = []
    })

    test('renders nothing', () => {
      createWrapper()
      expect(wrapper.html()).toEqual(``)
    })
  })

  describe('when keySoftware prop is set to one key software item with 1 version', () => {
    beforeEach(() => {
      propsData.keySoftware = [
        {
          name: 'Test Key Software',
          versions: ['1.2.3'],
        },
      ]
    })

    test('renders 1 badge containing 1 version', () => {
      createWrapper()
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('when keySoftware prop is set to one key software item with multiple versions', () => {
    beforeEach(() => {
      propsData.keySoftware = [
        {
          name: 'Test Key Software',
          versions: ['1.0.0', '1.1.0', '1.2.0'],
        },
      ]
    })

    test('renders 1 badge containing all the versions', () => {
      createWrapper()
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('when keySoftware prop is set to multiple key software items', () => {
    beforeEach(() => {
      propsData.keySoftware = [
        {
          name: 'Test Key Software 1',
          versions: ['1.0.0', '1.1.0', '1.2.0'],
        },
        {
          name: 'Test Key Software 2',
          versions: ['2.0.0', '2.1.0', '2.2.0'],
        },
        {
          name: 'Test Key Software 3',
          versions: ['3.0.0', '3.1.0', '3.2.0'],
        },
      ]
    })

    test('renders 3 badges', () => {
      createWrapper()
      expect(wrapper.html()).toMatchSnapshot()
    })
  })
})
