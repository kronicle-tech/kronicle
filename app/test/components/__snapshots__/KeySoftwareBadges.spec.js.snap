// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KeySoftwareBadges when keySoftware prop is set to multiple key software items renders 3 badges 1`] = `
"<div class=\\"h5\\">
  <b-badge variant=\\"info\\" class=\\"mr-2 mb-2\\">
    Test Key Software 1 1.0.0, 1.1.0, 1.2.0
  </b-badge>
  <b-badge variant=\\"info\\" class=\\"mr-2 mb-2\\">
    Test Key Software 2 2.0.0, 2.1.0, 2.2.0
  </b-badge>
  <b-badge variant=\\"info\\" class=\\"mr-2 mb-2\\">
    Test Key Software 3 3.0.0, 3.1.0, 3.2.0
  </b-badge>
</div>"
`;

exports[`KeySoftwareBadges when keySoftware prop is set to one key software item with 1 version renders 1 badge containing 1 version 1`] = `
"<div class=\\"h5\\">
  <b-badge variant=\\"info\\" class=\\"mr-2 mb-2\\">
    Test Key Software 1.2.3
  </b-badge>
</div>"
`;

exports[`KeySoftwareBadges when keySoftware prop is set to one key software item with multiple versions renders 1 badge containing all the versions 1`] = `
"<div class=\\"h5\\">
  <b-badge variant=\\"info\\" class=\\"mr-2 mb-2\\">
    Test Key Software 1.0.0, 1.1.0, 1.2.0
  </b-badge>
</div>"
`;
